test/integration/revocation_test.go

```diff
@@ -32,109 +33,163 @@ func isPrecert(cert *x509.Certificate) bool {
-			name:      "revocation by certificate key",
-			revokeKey: certKey,
-		},
-		{
-			name:      "revocation by owner account key",
-			revokeKey: c.Account.PrivateKey,
-		},
-		{
-			name:         "equivalently authorized account key",
-			revokeClient: otherAccount,
-			revokeKey:    otherAccount.Account.PrivateKey,
-			domain:       preAuthDomain,
-		},
+	type authMethod string
+	var (
+		byAccount authMethod = "byAccount"
+		byAuth    authMethod = "byAuth"
+		byKey     authMethod = "byKey"
+	)
+
+	type certKind string
+	var (
+		finalcert certKind = "cert"
+		precert   certKind = "precert"
+	)
+
+	type testCase struct {
+		method      authMethod
+		reason      int
+		kind        certKind
+		expectError bool
 	}
 
-	for _, tc := range testCases {
-		t.Run(tc.name, func(t *testing.T) {
-			// If the test case didn't specify a domain make one up randomly
-			if tc.domain == "" {
-				tc.domain = random_domain()
-			}
-			// If the test case didn't specify a different client to use for
-			// revocation use c.
-			if tc.revokeClient == nil {
-				tc.revokeClient = c
+	testCases := make([]testCase, 0)
+	for _, kind := range []certKind{precert, finalcert} {
+		for _, reason := range []int{ocsp.Unspecified, ocsp.KeyCompromise} {
+			for _, method := range []authMethod{byAccount, byAuth, byKey} {
+				testCases = append(testCases, testCase{
+					method: method,
+					reason: reason,
+					kind:   kind,
+					// We expect an error only for KeyComprommise requests that use auth
+					// methods other than using the certificate key itself.
+					expectError: (reason == ocsp.KeyCompromise) && (method != byKey),
+				})
 			}
+		}
+	}
 
-			// Make sure the ct-test-srv will reject issuance for the domain
-			err := ctAddRejectHost(tc.domain)
-			test.AssertNotError(t, err, "adding ct-test-srv reject host")
-
-			// Issue a certificate for the name using the `c` client. It should fail
-			// because not enough SCTs can be collected, leaving a precert without
-			// a matching final cert.
-			_, err = authAndIssue(c, certKey, []string{tc.domain})
-			test.AssertError(t, err, "expected error from authAndIssue, was nil")
-			if !strings.Contains(err.Error(), "urn:ietf:params:acme:error:serverInternal") ||
-				!strings.Contains(err.Error(), "SCT embedding") {
-				t.Fatal(err)
-			}
+	for _, tc := range testCases {
+		name := fmt.Sprintf("%s_%d_%s", tc.kind, tc.reason, tc.method)
+		t.Run(name, func(t *testing.T) {
+			issueClient, err := makeClient()
+			test.AssertNotError(t, err, "creating acme client")
+
+			certKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+			test.AssertNotError(t, err, "creating random cert key")
+
+			domain := random_domain()
 
-			// Try to find a precertificate matching the domain from one of the
-			// configured ct-test-srv instances.
-			cert, err := ctFindRejection([]string{tc.domain})
-			if err != nil || cert == nil {
-				t.Fatalf("couldn't find rejected precert for %q", tc.domain)
+			if tc.kind == precert {
```

Comment:
Empty `if` stanza?

Reply:
Oops, good catch! This was left over from when I was using if/else instead of the `switch tc.kind` that I ended up with below.